rust/src/dns/dns.rs

```diff
@@ -773,13 +776,17 @@ pub unsafe extern "C" fn rs_dns_parse_response_tcp(flow: *const core::Flow,
 }
 
 #[no_mangle]
-pub extern "C" fn rs_dns_tx_get_alstate_progress(_tx: *mut std::os::raw::c_void,
+pub unsafe extern "C" fn rs_dns_tx_get_alstate_progress(_tx: *mut std::os::raw::c_void,
                                                  _direction: u8)
                                                  -> std::os::raw::c_int
 {
     // This is a stateless parser, just the existence of a transaction
-    // means its complete.
+    // means its complete. However this is only true for UDP and not TCP
+    let tx = cast_pointer!(_tx, DNSTransaction);
     SCLogDebug!("rs_dns_tx_get_alstate_progress");
+    if tx.is_tcp && _direction & core::STREAM_TOCLIENT != 0 && !tx.response.is_some() {
+        return 0;
+    }
```

Comment:
I think this needs to be more complete, then fuzzing appears to pass.  See https://github.com/OISF/suricata/pull/7825/files#diff-fe45d440e3be98030915c1946205744d6a8cd791fad6223d1926c67705a91be4R787 for what I mean, feel free to integrate that into your PR.

Reply:
I integate the change and it works also without the tcp flag.

However some logging output is missing when analyzing udp dns packages. It no longer logs the dns traffic. Output like the following is missing with the updated patch:

{"timestamp":"2018-10-21T01:41:56.742793+0000","flow_id":1939439079413129,"pcap_cnt":1,"event_type":"dns","src_ip":"192.168.1.18","src_port":1027,"dest_ip":"192.168.1.1","dest_port":53,"proto":"UDP","pkt_src":"wire/pcap","dns":{"type":"query","id":17024,"rrname":"time.windows.com","rrtype":"A","tx_id":0}}
{"timestamp":"2018-10-21T01:41:56.743525+0000","flow_id":1939439079413129,"pcap_cnt":2,"event_type":"dns","src_ip":"192.168.1.18","src_port":1027,"dest_ip":"192.168.1.1","dest_port":53,"proto":"UDP","pkt_src":"wire/pcap","dns":{"version":2,"type":"answer","id":17024,"flags":"8180","qr":true,"rd":true,"ra":true,"rrname":"time.windows.com","rrtype":"A","rcode":"NOERROR","answers":[{"rrname":"time.windows.com","rrtype":"CNAME","ttl":3600,"rdata":"time.microsoft.akadns.net"}],"grouped":{"CNAME":["time.microsoft.akadns.net"]}}}

Any idea why that is happening?

Reply:
I think it is the bug described in https://github.com/OISF/suricata/pull/7693#issuecomment-1226958253 which needs a bigger fix, not only for DNS.

A suricata-verify pull request would be nice to have all these expected behaviors

Reply:
It seems that I was wrong. The logs are created when including https://github.com/OISF/suricata/pull/7825/files#diff-fe45d440e3be98030915c1946205744d6a8cd791fad6223d1926c67705a91be4R787.
Seems I had not used the correct patch when I tested. I will update my pull request.

When I remove the tcp flag the logs no longer work.

Reply:
> When I remove the tcp flag the logs no longer work.

Ok.  It is ok to leave the tcp check in there.  I think this gets us to having dns behave better with other protocols, but as mentioned, there are some other issues at play here that likely need some looking at outside just dns.

Do you plan to do a new PR with this change?

Reply:
> Do you plan to do a new PR with this change?

The new pull request can be found here https://github.com/OISF/suricata/pull/7835